### PR TITLE
MAINT: Refactoring filtering functions

### LINKED ISSUE #15: ENH+REVIEW: Enhancement for filter reads actions (for single-end and paired-end reads)
The code and execution of the actions `filter_paired_end_reads` and `filter_single_end_reads` needs additional review to detect potential enhancements or bugs.

The `filter_paired_end_reads` action aligns paired-end reads to a reference genome or index using Minimap2, processes the alignment to filter out reads based on the specified percentage identity and either keeps the mapped or the unmapped sequences based on what the user specifies. It modifies flags for paired and unmapped reads appropriately in order to be recognized by the samtools fastq command which converts the filtered alignments into separate FASTQ files for each read in the pair. This allows users to handle paired-end sequencing data specifically, ensuring the reads are appropriately paired and filtered.

The `filter_single_end_reads` function performs a similar process but for single-end reads. It aligns these reads to a reference genome, filters the alignments based on identity and either keeps the mapped or the unmapped sequences based on what the user specifies. Then it converts the filtered alignments into initial FASTQ format. This function is tailored for single-end reads, focusing on filtering individual reads without considering pair relationships.

### PR DESCRIPTION
**Pull request by ChristosMatzoros**
This PR enhances action `filter-reads` that closes #15 .

### Set up the environment
 ```shell
mamba create -n q2-minimap2 -c conda-forge -c bioconda -c https://packages.qiime2.org/qiime2/2024.2/shotgun/passed/ -c defaults q2cli q2-types q2-feature-classifier minimap2 bs4 samtools
```

###  Activate q2-minimap2 environment
```shell
conda activate q2-minimap2
```

### Clone the repo and checkout the PR branch
```bash
git clone https://github.com/bokulich-lab/q2-minimap2.git
cd q2-minimap2
git fetch origin pull/26/head:pr-16
git checkout pr-16
pip install -e .
```

1. Download Datasets
[here](https://polybox.ethz.ch/index.php/s/0hduvR3Q7jeiC26)

2. Test it out!
```bash
# filtering mapped single-end reads by mapping using minimap2 index
qiime minimap2 filter-reads --i-query-reads single-end-reads.qza --i-index-database hum_ref_mn2_index.qza  --o-filtered-query-reads mapped-se.qza --verbose
```
```bash
# filtering mapped single-end reads by mapping using a reference sequence
qiime minimap2 filter-reads --i-query-reads single-end-reads.qza --i-reference-reads human_reference.qza  --o-filtered-query-reads mapped-se-ref.qza --verbose
```
```bash
# filtering mapped paired-end reads by mapping using minimap2 index
qiime minimap2 filter-reads --i-query-reads paired-end-reads.qza --i-index-database hum_ref_mn2_index.qza  --o-filtered-query-reads mapped-pe.qza --verbose
```
```bash
# filtering mapped single-end reads by mapping using minimap2 index
# keep as mapped only the sequences with percentage identity to the reference over 85%
qiime minimap2 filter-reads --i-query-reads single-end-reads.qza --i-index-database hum_ref_mn2_index.qza --p-min-per-identity 0.85 --o-filtered-query-reads mapped-se-mpid85.qza --verbose
```
```bash
# filtering unmapped single-end reads by mapping using minimap2 index
qiime minimap2 filter-reads --i-query-reads single-end-reads.qza --i-index-database hum_ref_mn2_index.qza --p-keep unmapped  --o-filtered-query-reads unmapped-se.qza --verbose
```
```bash


```

### CODE REVIEW

**Comment by ChristosMatzoros** [comment]
This PR aims to explore any additional enhancements for the two filtering actions, `filter_paired_end_reads` and `filter_single_end_reads`, and to review the existing codebase for both.

**Comment by VinzentRisch** [comment]
add_mapped_paired_read_flags
- add test 
- maybe add a docstring

**Comment by ChristosMatzoros** [comment]
Hey @VinzentRisch,

I made the changes you suggested and added the tests. I also merged `add_mapped_paired_read_flags()` and `process_paired_unmapped_flags()` functions into one simpler function (`process_paired_sam_flags()`). Please have a look and let me know. Thank you

**Comment by ChristosMatzoros** [comment]
Dear @VinzentRisch,

I added a necessary test for paired-end reads, and now I think its ready. Please have a look and let me know if something else is needed.  

Thanks

**Review by VinzentRisch — CHANGES_REQUESTED** [review]
Hey :) see some changes below.

File: q2_minimap2/_filtering_utils.py [hunk 1]
```diff
@@ -76,61 +66,55 @@ def get_alignment_length(cigar):
...
                 if not (flag & 0x4) and not (flag & 0x100):
-                    # If the minimum identity percentage is provided then accept the
-                    # mapped read if its identity percentage is equal or exceeds
-                    # the minimum identity percentage
-                    if (min_per_identity is None) or (
-                        identity_percentage >= min_per_identity
+                    if (
+                        min_per_identity is None
+                        or identity_percentage >= min_per_identity
                     ):
```
Comment (VinzentRisch) [inline q2_minimap2/_filtering_utils.py:102]:
```suggestion
                       if not min_per_identity or identity_percentage >= min_per_identity:
```
This fits on one line. I think it is easier to read

**Review by VinzentRisch — APPROVED** [review]
Hey looks good :)

**Pull request merged**